core/domain/classifier_services.py

```diff
@@ -100,8 +101,38 @@ def classify_string_classifier_rule(state, normalized_answer):
             training_examples.extend([
                 [doc, [str(answer_group_index)]]
                 for doc in classifier_rule_spec.inputs['training_data']])
-    if len(training_examples) > 0:
+
+    exploration = exp_services.get_exploration_by_id(exp_id)
+    state_obj = exploration.states[state_name]
+    algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING['TextInput']
+    if state_obj.classifier_model_id:
+        classifier = get_classifier_by_id(state_obj.classifier_model_id)
+        sc_dict = classifier.cached_classifier_data
+        sc = classifier_registry.Registry.get_classifier_by_algorithm_id(
+            algorithm_id)
+        sc.from_dict(sc_dict)
+
+    else:
+        sc = classifier_registry.Registry.get_classifier_by_algorithm_id(
+            algorithm_id)
         sc.train(training_examples)
+        cached_classifier_data = sc.to_dict()
+        algorithm_version = feconf.CLASSIFIER_VERSION_MAPPING[algorithm_id]
+        classifier = classifier_domain.Classifier(
+            '0', exploration.id, exp_version, state_name, algorithm_id,
+            cached_classifier_data, algorithm_version)
+        classifier_id = save_classifier(classifier)
+        #state.classifier_model_id = classifier_id
+        change_list = [{
+            'cmd': 'edit_state_property',
+            'state_name': state_name,
+            'property_name': 'classifier_model_id',
+            'new_value': classifier_id
+        }]
+        exp_services.update_exploration(
```

Comment:
Ah, I see - we need to update the classifier_id for the state if a new classifier is created. So, in this case I think the creator is the owner who triggered the training task and we will have to get his ID instead of using this dummy owner_id, right?

FYI, a lot of things from the old code have been left incomplete or they are just hanging around which could have side effects (like this dummy owner id). @pranavsid98 if you seem dubious about anything, please feel free to raise a flag.